gprimitives/src/lib.rs

```diff
@@ -0,0 +1,184 @@
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Gear primitive types.
+
+#![no_std]
+#![warn(missing_docs)]
+#![doc(html_logo_url = "https://docs.gear.rs/logo.svg")]
+#![doc(html_favicon_url = "https://gear-tech.io/favicons/favicon.ico")]
+
+pub use gear_ss58::Ss58Address;
+
+pub mod macros;
+
+use core::{
+    fmt,
+    str::{self, FromStr},
+};
+use derive_more::{AsMut, AsRef, Display, From, Into};
+use gear_ss58::RawSs58Address;
+#[cfg(feature = "codec")]
+use {
+    primitive_types::H256,
+    scale_info::{
+        scale::{self, Decode, Encode, MaxEncodedLen},
+        TypeInfo,
+    },
+};
+
+/// The error type returned when conversion fails.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Display)]
+pub enum ConversionError {
+    /// Invalid slice length.
+    #[display(fmt = "Slice should be 32 length")]
+    InvalidSliceLength,
+    /// Invalid hex string.
+    #[display(fmt = "Invalid hex string")]
+    InvalidHexString,
+    /// Invalid SS58 address.
+    #[display(fmt = "Invalid SS58 address")]
+    InvalidSs58Address,
+    /// SS58 encoding failed.
+    #[display(fmt = "SS58 encoding failed")]
+    Ss58Encode,
+}
+
+/// Message handle.
+///
+/// Gear allows users and programs to interact with other users and programs via
+/// messages. Message creation consists of the following parts: message
+/// initialization, filling the message with payload (can be gradual), and
+/// message sending.
+#[repr(transparent)]
+#[derive(Clone, Copy, Debug, PartialEq, Eq, From, Into)]
+#[cfg_attr(feature = "codec", derive(TypeInfo, Encode, Decode, MaxEncodedLen), codec(crate = scale))]
+pub struct MessageHandle(u32);
+
+/// Program (actor) identifier.
+///
+/// Gear allows user and program interactions via messages. Source and target
+/// program as well as user are represented by 256-bit identifier `ActorId`
+/// struct. The source `ActorId` for a message being processed can be obtained
+/// using `msg::source()` function. Also, each send function has a target
+/// `ActorId` as one of the arguments.
+#[derive(Clone, Copy, Default, Hash, Ord, PartialEq, PartialOrd, Eq, From, Into, AsRef, AsMut)]
+#[as_ref(forward)]
+#[as_mut(forward)]
+#[cfg_attr(feature = "codec", derive(TypeInfo, Encode, Decode, MaxEncodedLen), codec(crate = scale))]
+pub struct ActorId([u8; 32]);
+
+macros::impl_primitive!(new zero into_bytes from_u64 from_h256 try_from_slice display debug, ActorId);
```

Comment:
Looks like there is no need to implement different sets of functions for each ID

Reply:
What's the point of `ReservationId::zero()` then? Some methods are useless because you can only get this type using `ReservationId::reserve()`.